csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,290 @@
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Record all ValGroups that are producers of those in abstract_tensor_.
+  //! This lets us find starting IterDomains for scheduling concrete's loop
+  //! domain.
+  //! TODO: this should probably use a new more general utility that resembles
+  //! StmtSort::getExprsTo in val_graph_visitor.h
+  void findScheduledValGroups() {
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_tensor_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups_.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (const ValGroup& vg_inp : graph_->inputGroups(eg)) {
+          vg_stack.push(vg_inp);
+        }
+      }
+    }
+  }
+
+  //! Look at all ValGroups that are producers of abstract_tensor_.domain and
+  //! map any of those containing loop IterDomains in tv to those loop
+  //! IterDomains. These provide the starting points for scheduling.
+  GroupIdMap mapScheduledGroupsToLoopIterDomains(TensorView* tv) const {
+    GroupIdMap tv_ids;
+
+    // Look up each loop IterDomain in tv and assert that it is in
+    // scheduled_val_groups_, then map it.
+    for (IterDomain* loop_id : tv->getLoopDomain()) {
+      ValGroup vg = graph_->toGroup(loop_id);
+      // NOTE: this check guarantees that we will only ever _append_ transforms
+      // to the loop domain of tv. If instead we searched for producer IDs in
+      // the root->loop path of tv, then we could wind up silently erasing
+      // some pre-existing transforms.
+      NVF_CHECK(
+          scheduled_val_groups_.count(vg) > 0,
+          "Found loop IterDomain ",
+          loop_id->toString(),
+          " that is not described by abstract tensor.");
+      tv_ids.emplace(vg, loop_id);
+    }
+
+    return tv_ids;
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id, GroupIdMap& tv_ids) {
+    ValGroup g = abstractIdToValGroup(abs_id);
+
+    // This holds ValGroups that we cannot compute from tv's root. When
+    // we detect that any inputs of an ExprGroup are in these groups, we know
+    // that we cannot compute that ExprGroup so we should skip it.
+    //
+    // For example, suppose we had
+    //   tv:
+    //     logical/loop: iS0{i0} iS1{i1} iS2{i2}
+    //
+    //   graph:
+    //     ValGroup 0: iS0
+    //     ValGroup 1: iS1
+    //     ValGroup 2: iS2 iS6
+    //     ValGroup 3: iS3
+    //     ValGroup 4: iS4
+    //     ValGroup 5: iS5
+    //     ValGroup 6: iS7
+    //     ValGroup 7: iS8
+    //     ExprGroup 0:
+    //       iS3 = merge(iS0, iS1)
+    //     ExprGroup 1:
+    //       iS4, iS5 = split(iS3, 16)
+    //     ExprGroup 2:
+    //       iS8 = merge(iS6, iS7)
```

Comment:
Where are `iS6` and `iS7` defined?

Reply:
They are defined elsewhere and could be associated with a `TensorView` or not. For example, `AbstractTensor::split` and `AbstractTensor::merge` will create such `IterDomain`s in order to create an `ExprGroup`. I'll update the comment to try and clarify.

Reply:
I suspected that but wasn't clear if it's just a mistake. Can you please clarify that in the comment?

Reply:
I added this part to the comment: https://github.com/NVIDIA/Fuser/blob/7a51c9a9ee668ade6ccdba92453a34a48729ce6a/csrc/abstract_tensor_schedule.cpp#L172-L185